### DISCUSSION TITLE
How to avoid 504 Gateway Timeout errors

### DISCUSSION THREAD

**Discussion by jonathan-marcus** [fastapi/fastapi#4365] [Questions] [answered]
**Problem**: I'm deploying my FastAPI server behind a proxy with a 30 second timeout. Most of my API endpoints return quickly but a few of them hit external services that can sometimes take a while, so users occasionally get a 504 Gateway Timeout error.

**Possible solution**: I can't change the 30 second limit, but if I send a single byte then the proxy knows the server's not dead and it disables the timeout. I'd like my affected endpoints to emit a space character right at the beginning and then return the rest of the response as normal.

**Question**: Is there a way I can wrap my endpoint methods to make this happen? I could change each method to use a StreamingResponse, and change the internals to use an async method, but that feels very clunky. It would be nice if I could add something to my `@app.get` decorator call instead.

Any ideas on how to make this happen?

**Comment by bluebrown** [comment, marked as answer]
The proper way is to switch to an asynchronous request/response model. If you need longer than 30 seconds to reply, you better reply with a 202 immediately and let the client know where they can monitor the status of the task.

**Reply by jonathan-marcus** [reply]
I see where you're coming from, but that would be a pretty radical redesign just to fix a pretty rare bug.

For context, this is an internal app whose users are primarily automated scripts that are very tolerant of long response times, but not tolerant of errors. Waiting for the response is much easier for these scripts to handle versus forcing the client to do a multistep process (request, poll, poll, poll, ...).

I'd much rather emit a single whitespace character at the beginning, which would require no change of client code.

**Comment by Almazishe** [comment]
Did you solve this problem? I think I have the same, in some endpoints which have Depends or maybe APScheduler add_job call, they mostly works but in some time they start giving out 504 Timeout. Or maybe this is not because of DB and APScheduler))
![Screenshot from 2022-05-27 08-40-10](https://user-images.githubusercontent.com/67875145/170620130-b817bab1-6211-4d80-8a59-109d02d266ae.png)
![Screenshot from 2022-05-27 08-38-56](https://user-images.githubusercontent.com/67875145/170620166-4305e8d4-6086-402e-b65e-266c41a236cc.png)

**Comment by binoytv9** [comment]
I tried below 2 things and worked for me

- Increase timeout in Gunicorn (https://stackoverflow.com/a/24305939/3949786)
- Increase timeout in Nginx (https://www.ibm.com/docs/en/fci/6.5.0?topic=tls-504-gateway-time-out-error)
